### PR TITLE
Add Post Image Upload Functionality

### PR DESCRIPTION
**Pull request by robin-macpherson**
## Description

**Issue:** *Users should be able to upload images to their posts when writing a new post.*

**Feature Set:**

- Upload post 'HEADLINE' image, preserving the default image if the user doesn't upload their own
- Show an actual preview of how the final post header will look!
- Ability to upload a different image and also reset back to the default
- Auto-saving even the image!

## Full Flow

image-upload-feature-preview

## Screenshot

image

## Subtasks

- [x] I have added this PR to the Journaly Kanban project ✅
- [x] Scaffold everything out
- [x] Figure out properly creating the `postId` relationship
- [x] Update graphql mutation
- [x] Test all the things
- [x] Dance, don't panic
- [x] sdgkjsdk;glghskldjlgksd THIS IS SO FREAKIN' SWEET!!!!!

### CODE REVIEW

File: frontend/pages/dashboard/new-post.tsx [hunk 1]
```diff
@@ -56,8 +94,10 @@ const NewPostPage: NextPage = () => {
     .concat(languagesNative.map((x) => x.language))
 
   const createNewPost = (status: PostStatusType) => {
+    const images = []
+    images.push(image)
     createPost({
-      variables: { title, body, status, languageId: langId },
+      variables: { title, body, status, languageId: langId, images },
```
Comment (Lanny) [inline frontend/pages/dashboard/new-post.tsx:100]: Just to cut down on line noise you can do:

```suggestion
      variables: { title, body, status, languageId: langId, images: [image] },
```

and get rid of lines 97-98

**Comment by Lanny** [comment]
oop, hold, vercel deployment didn't go through. We should probably make that a required check

File: frontend/pages/dashboard/new-post.tsx [hunk 2]
```diff
@@ -87,6 +139,35 @@ const NewPostPage: NextPage = () => {
...
+          <div className="header-preview-container">
+            <PostHeader
+              postTitle={title}
+              postStatus={PostStatusType.Published}
+              publishDate={new Date().toISOString()}
+              authorName={currentUser?.name || 'anonymous'}
+              postImage={image.largeSize || '/images/samples/sample-post-img.jpg'}
+            >
+              <div className="header-preview-options">
+                <Button
```
Comment (nlitwin) [inline frontend/pages/dashboard/new-post.tsx:151]: I saw in the PR description GIF that we could potentially benefit from adding a value for the `loading` prop here on `<Button />`. We could add something like `const [uploadingImage, setUploadingImage] = useState(false)` and set it to true when uploading the file and then use

File: frontend/pages/dashboard/new-post.tsx [hunk 3]
```diff
@@ -87,6 +139,35 @@ const NewPostPage: NextPage = () => {
             onChange={setLangId}
           />
 
+          <div className="header-preview-container">
+            <PostHeader
+              postTitle={title}
+              postStatus={PostStatusType.Published}
+              publishDate={new Date().toISOString()}
+              authorName={currentUser?.name || 'anonymous'}
```
Comment (nlitwin) [inline frontend/pages/dashboard/new-post.tsx:147]: When would we see the `anonymous` value?

File: frontend/pages/dashboard/new-post.tsx (hunk 2 above)
Comment (robin-macpherson) [reply frontend/pages/dashboard/new-post.tsx:151]: Yep yep!! I was actually wanting to add that in right after work! Wasn't 100% sure how best to implement it yet so I'll try your suggestion!

File: frontend/pages/dashboard/new-post.tsx (hunk 3 above)
Comment (robin-macpherson) [reply frontend/pages/dashboard/new-post.tsx:147]: We'll never see it, this was mostly just to satisfy TypeScript 😊

**Comment by Lanny** [comment]
re deploy failure, looks like a TS error:

```

12:59:30.255 | ERROR in /vercel/24ac448a/pages/dashboard/new-post.tsx(155,23):
-- | --
12:59:30.255 | 155:23 Object is possibly 'null'.
12:59:30.255 | 153 \|                     e.preventDefault()
12:59:30.255 | 154 \|                     if (fileInput && fileInput.current) {
12:59:30.255 | > 155 \|                       fileInput.current.click()
12:59:30.256 | \|                       ^
12:59:30.256 | 156 \|                     }
12:59:30.256 | 157 \|                   }}
12:59:30.256 | 158 \|                   className="image-upload-btn"
12:59:30.256 | > Build error occurred
12:59:30.257 | Error: > Build failed because of webpack errors
12:59:30.257 | at build (/vercel/24ac448a/node_modules/next/dist/build/index.js:13:900)
```

I imagine `npm run build` will reproduce it but not sure. I think you need to provide `HTMLInputElement` as the type argument to `useRef`

**Comment by robin-macpherson** [comment]
> re deploy failure, looks like a TS error:
> 
> ```
> 
> 12:59:30.255 | ERROR in /vercel/24ac448a/pages/dashboard/new-post.tsx(155,23):
> -- | --
> 12:59:30.255 | 155:23 Object is possibly 'null'.
> 12:59:30.255 | 153 \|                     e.preventDefault()
> 12:59:30.255 | 154 \|                     if (fileInput && fileInput.current) {
> 12:59:30.255 | > 155 \|                       fileInput.current.click()
> 12:59:30.256 | \|                       ^
> 12:59:30.256 | 156 \|                     }
> 12:59:30.256 | 157 \|                   }}
> 12:59:30.256 | 158 \|                   className="image-upload-btn"
> 12:59:30.256 | > Build error occurred
> 12:59:30.257 | Error: > Build failed because of webpack errors
> 12:59:30.257 | at build (/vercel/24ac448a/node_modules/next/dist/build/index.js:13:900)
> ```
> 
> I imagine `npm run build` will reproduce it but not sure. I think you need to provide `HTMLInputElement` as the type argument to `useRef`

haha just DM'd you about that I think the exact second you left this comment 😂

File: frontend/pages/dashboard/new-post.tsx (hunk 2 above)
Comment (robin-macpherson) [reply frontend/pages/dashboard/new-post.tsx:155]: Done!!! ✅

File: frontend/pages/dashboard/new-post.tsx (hunk 1 above)
Comment (robin-macpherson) [reply frontend/pages/dashboard/new-post.tsx]: Wonderful wee suggestion! Done!

**Pull request merged**